src/weight_windows.cpp

```diff
@@ -0,0 +1,265 @@
+#include "openmc/weight_windows.h"
+
+#include "openmc/error.h"
+#include "openmc/file_utils.h"
+#include "openmc/hdf5_interface.h"
+#include "openmc/particle.h"
+#include "openmc/particle_data.h"
+#include "openmc/search.h"
+#include "openmc/xml_interface.h"
+
+#include <fmt/core.h>
+#include <gsl/gsl-lite.hpp>
+
+namespace openmc {
+
+//==============================================================================
+// Global variables
+//==============================================================================
+
+namespace variance_reduction {
+
+std::unordered_map<int32_t, int32_t> ww_map;
+openmc::vector<unique_ptr<WeightWindows>> weight_windows;
+
+} // namespace variance_reduction
+
+//==============================================================================
+// Non-member functions
+//==============================================================================
+
+void apply_weight_windows(Particle& p)
+{
+  // skip dead or no energy
+  if (p.E() <= 0 || !p.alive())
+    return;
+
+  bool in_domain = false;
+  // TODO: this is a linear search - should do something more clever
+  WeightWindow weight_window;
+  for (const auto& ww : variance_reduction::weight_windows) {
+    weight_window = ww->get_weight_window(p);
+    if (weight_window)
+      break;
+  }
+  // particle is not in any of the ww domains, do nothing
+  if (!weight_window)
+    return;
+
+  // get the paramters
+  double weight = p.wgt();
+  p.wgt_last() = weight;
+
+  // first check to see if particle should be killed for weight cutoff
+  if (p.wgt() < weight_window.weight_cutoff) {
+    p.alive() = false;
+    p.wgt() = 0.0;
+    return;
+  }
+
+  // if particle's weight is above the weight window split until they are within
+  // the window
+  if (weight > weight_window.upper_weight) {
+    // do not further split the particle if above the limit
+    if (p.n_split() >= settings::max_splits)
+      return;
+
+    double n_split = std::ceil(weight / weight_window.upper_weight);
+    double max_split = weight_window.max_split;
+    n_split = std::min(n_split, max_split);
+
+    p.n_split() += n_split;
+
+    // Create secondaries and divide weight among all particles
+    int i_split = std::round(n_split);
+    for (int l = 0; l < i_split - 1; l++) {
+      p.create_secondary(weight / n_split, p.u(), p.E(), p.type());
+    }
+    // TODO: maybe weight should be weight - sum of child weight
```

Comment:
Remove TODO? Maybe just make a note saying we're assuming even weight splitting here. Using the sum of child weights could possibly accrue a small amount of floating point error.